Making Circular Option Picker a `listbox`

andrewhayward opened the pull request:
## What?
As per #35292, This patch modifies [`CircularOptionPicker`](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/circular-option-picker) to become a [`listbox`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role), but using [the `Composite` package](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/composite/index.js). Additionally, it updates the various components that use `CircularOptionPicker`.

## Why?
Currently, keyboard interaction with `CircularOptionPicker`s is difficult, as each option presents as an individual tab stop. By changing the component to behave as a `listbox`, the entire control becomes a single tab stop, with individual colour options accessed using arrow keys.

## How?
The `CircularOptionPicker` has been partially rebuilt using the various `Composite` elements, which reduces any complex behavioural additions on our part.

## Testing Instructions
Nothing should visually change, as the underlying base components are still being used.

### Testing Instructions for Keyboard
Every use of the `CircularOptionPicker` ([`ColorPalette`](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/color-palette), for example) should now present as a single tab stop, with options being picked using arrow keys.

## Additional Notes
 - ~~Initial values aren't currently picked up correctly.~~

## ✍️ Dev note

To improve `CircularOptionPicker`'s semantics and keyboard navigation, the component has been tweaked to render and behave as a `listbox` by default. This change also causes the component to become a single tab stop, with the individual color options accessed using arrow keys.

In the (few) instances in which it makes sense for `CircularOptionPicker` to still render as a list of individual buttons,  consumers of the component can use the `asButtons` prop to switch back to the legacy behavior.

Commit:
Making Circular Option Picker a `listbox`

Using the `Composite` component to make the Circular Option Picker present as a listbox.

mirka commented on packages/components/src/circular-option-picker/types.ts:
```diff
@@ -36,6 +36,11 @@ export type CircularOptionPickerProps = {
 	children?: ReactNode;
+	/**
+	 * Whether the keyboard interaction should wrap around.
+	 * Defaults to `true`.
```
```suggestion
	 *
	 * @default true
```

mirka commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -58,6 +73,34 @@ export function Option( {
+	// This is unlikely to happen, but on the off-chance that we've ended up
+	// with a list of `OptionGroup`s, we will just return those instead.
+	if ( Array.isArray( options ) && options[ 0 ] instanceof OptionGroup ) {
+		return <>options</>;
```
Let's make sure this works visually, and also add a unit test. I don't think we can use `instanceof` in this way.

```suggestion
	if ( Array.isArray( options ) && options[ 0 ] instanceof OptionGroup ) {
		return <>{ options }</>;
```

mirka commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -152,28 +195,37 @@ export function ButtonAction( {
-			) }
-		</div>
+			<CircularOptionPickerContext.Provider value={ compositeState }>
+				{ Array.isArray( options ) ? (
```
`options` is not necessarily an array, as can be [seen](https://github.com/WordPress/gutenberg/blob/5ba748472b3e481e89ec3ac4c4e215031cc1eb3c/packages/components/src/circular-option-picker/stories/index.tsx#L109C1-L110) in the [default Storybook story](https://wordpress.github.io/gutenberg/?path=/story/components-circularoptionpicker--default):

mirka commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -105,19 +106,21 @@ function MultiplePalettes( {
 	return (
 		<VStack spacing={ 3 } className={ className }>
 			{ colors.map( ( { name, colors: colorPalette }, index ) => {
+				const id = `color-palette-${ instanceId }-${ index }`;
```
Can we use the `prefix` argument of [`useInstanceId()`](https://github.com/WordPress/gutenberg/blob/5ba748472b3e481e89ec3ac4c4e215031cc1eb3c/packages/compose/src/hooks/use-instance-id/index.ts#L40-L48) to simplify this?

mirka commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -91,10 +92,10 @@ function SinglePalette( {
 
 	return (
-		<CircularOptionPicker
+		<CircularOptionPicker.OptionGroup
```
I'm unsure about modeling the multiple palette case as `group`s in a single `listbox`. From reading the [WAI guidance](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/), I think a `listbox` is assumed to be one-dimensional (i.e. a list not a grid), with a `aria-orientation` to dictate which arrow keys to use.

What do you think about dropping the `group` construct and modeling them as separate listboxes?

ciampo replied:
> What do you think about dropping the group construct and modeling them as separate listboxes?

After interacting with the `ColorPalette` storybook demo with multiple origins, I also got the feeling that having each `OptionGroup` as a separate tab stop feels better.

ciampo replied:
After further discussion, we've decided that we're going to keep the current behaviour, and iterate if/when necessary in a future PR.

mirka commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -152,28 +195,37 @@ export function ButtonAction( {
 function CircularOptionPicker( props: CircularOptionPickerProps ) {
-	const { actions, className, options, children } = props;
+	const { actions, className, options, children, loop = true } = props;
+	const compositeState = useCompositeState( { loop } );
```
I'm actually not up to date on this, but is there a reason why we're using `useCompositeState` instead of [`useCompositeStore`](https://github.com/ariakit/ariakit/blob/c30e6a67bebcde1b2bff9a8183e1f098259a14c7/packages/ariakit-react-core/src/composite/composite.tsx#L438-L449)?

andrewhayward replied:
We're still using `reakit` under the hood for this, which is where `useCompositeState` is coming from. As and when we move `composite` to `ariakit` we'll have to change that.

mirka commented on packages/components/src/circular-option-picker/index.tsx:
```diff
@@ -7,35 +7,50 @@ import classnames from 'classnames';
+		<CompositeItem
+			as={ Button }
+			className={ 'components-circular-option-picker__option' }
 			isPressed={ isSelected }
```
I think we need to change it from `aria-pressed` to `aria-selected`?

We may also want to make sure a consumer can add `aria-multiselectable` to the `listbox`.

mirka commented on packages/components/src/color-palette/stories/index.tsx:
```diff
@@ -33,9 +33,10 @@ export default meta;
 	...args
 } ) => {
-	const [ color, setColor ] = useState< string | undefined >();
+	const [ color, setColor ] = useState< string | undefined >( value );
```
(Marking this as TODO so we don't forget to address it)

ciampo replied:
Out of curiosity, what exactly was the TODO item related to this line of code?

ciampo replied:
@andrewhayward would you happen to know what needed to be done around this line?

mirka reviewed (commented):
Thank you for working on this big enhancement!

Based on my first-pass review, this approach looks viable to me. What do you think? If you agree, let's address the remaining issues and add some unit tests.

We also need to check for RTL support. I think we need to map `@wordpress/i18n`'s [`isRTL()`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-i18n/#isrtl) to the [Composite store's `rtl`](https://github.com/ariakit/ariakit/blob/c30e6a67bebcde1b2bff9a8183e1f098259a14c7/packages/ariakit-react-core/src/composite/composite-store.ts#L32).

andrewhayward committed:
Merge branch 'WordPress:trunk' into 35292/improve_color_palette_keyboard_accessiblity--listbox

Commit:
Various updates

– Adding support for RTL
- Adding support for preselected values
- Adjusting how groups behave
- Addressing some feedback

Commit:
Merge branch 'trunk' into 35292/improve_color_palette_keyboard_accessiblity--listbox

Commit:
Merge branch 'trunk' into 35290/improve_color_palette_keyboard_accessiblity--listbox

Commit:
Fixing tests for `core/cover`

Commit:
Fixing tests for `BorderControl`

Commit:
Fixing tests for `BorderControl`

Commit:
Adding default values for ARIA labels
Using `useEffect` to fix updates during render cycle

Commit:
Fixing merge conflict

Commit:
Removing redundant `ColorPalette` test

Commit:
Updating custom color tests

Commit:
Removing redundant snapshots

Commit:
Updating broken snapshot

Commit:
Merge branch 'trunk' into 35292/improve_color_palette_keyboard_accessiblity--listbox

Commit:
Updating CHANGELOG.md

andrewhayward commented:
@alexstine @joedolson could you give this a quick look, if you have a chance, and give me any thoughts/feedback? Thanks.

alexstine commented:
@andrewhayward Thanks for the ping. This is testing well for me in Chrome and Firefox.

Commit:
Fixing various E2E tests

Commit:
Updating test for new keyboard navigation

Commit:
Updating E2E tests

Commit:
Fixing navigation-colors E2E test

Commit:
Fixing E2E tests for cover and heading blocks

Commit:
Removing `option` constant

ciampo commented on packages/components/src/border-control/test/index.js:
```diff
@@ -75,6 +79,10 @@ const clickButton = ( name ) => {
 };
 
+const selectColorOption = ( name ) => {
+	fireEvent.click( getColorOption( name ) );
```
Another idea for a follow-up PR: refactor `BorderControl`'s unit tests from `fireEvent` to `testing-library/user-event`

andrewhayward replied:
See #54155.

ciampo commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -105,19 +106,21 @@ function MultiplePalettes( {
 	colors,
 	onChange,
 	value,
-	actions,
```
Just to make sure I'm following here, both `SinglePalette` and `MultiplePalettes` are not expecting an `actions` prop anymore because, later in the file, we're wrapping them in a `CircularOptionPicker` component (which accepts the `actions` itself). Is that correct?

andrewhayward replied:
Yes, that's correct. `CircularOptionPicker` now acts as a container for one or more palettes, so the palettes themselves no longer need to know about `actions`.

ciampo commented on packages/components/src/color-palette/index.tsx:
```diff
@@ -234,18 +237,26 @@ function UnforwardedColorPalette(
+		</CircularOptionPicker.ButtonAction>
+	);
+
+	let ariaProps: { 'aria-label': string } | { 'aria-labelledby': string };
+	if ( ariaLabel ) {
+		ariaProps = { 'aria-label': ariaLabel as string };
+	} else if ( ariaLabelledby ) {
+		ariaProps = { 'aria-labelledby': ariaLabelledby as string };
+	} else {
+		ariaProps = { 'aria-label': __( 'Custom color picker.' ) };
+	}
```
The `as string` type casts shouldn't be necessary:

```suggestion
	let ariaProps: { 'aria-label': string } | { 'aria-labelledby': string };
	if ( ariaLabel ) {
		ariaProps = { 'aria-label': ariaLabel };
	} else if ( ariaLabelledby ) {
		ariaProps = { 'aria-labelledby': ariaLabelledby };
	} else {
		ariaProps = { 'aria-label': __( 'Custom color picker.' ) };
	}
```

andrewhayward replied:
I wonder if I changed something somewhere, because I'm sure at one point it was complaining that the value was `string or undefined` and so couldn't be coerced into a `string`. Ho hum. Will fix.

ciampo commented on packages/components/src/color-palette/types.ts:
```diff
@@ -88,4 +88,16 @@ export type ColorPaletteProps = Pick< PaletteProps, 'onChange' > & {
 	 * @default false
 	 */
 	__experimentalIsRenderedInSidebar?: boolean;
+	/**
+	 * A label to identify the purpose of the control.
+	 *
+	 * @todo Either this or `aria-labelledby` should be required
+	 */
+	'aria-label'?: string;
+	/**
+	 * An ID of an element to provide a label for the control.
+	 *
+	 * @todo Either this or `aria-label` should be required
+	 */
+	'aria-labelledby'?: string;
```
If you want either prop `a` or prop `b` to be required, you should be able to do so by using the `never` keyword — see [this playground example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gWwXAdgBQE5zAZygXigG8AoKKAY0WRQC4odgMBLFAcwG4SBfLk0SFEzYAggCUIARwCuzDBAAmBWNVTDcUAGTEyUAIb1GLdl3IAjAPz0UEAG4QMXbiX7ho6gEISZcxcvhIalga2qTkelZQNvaOumaGTKycPC4C7sF4hOriUrLySgA+QsFeub4KfFQojFAA7szAABbZ9OqZOuRUgXRQAOTAEIy9ADS6Bn0DQzyVqDX1TZ6tGcphlKo9-YPAI3H0m1O8LgD0R1AAAsA4ALQQAB6QFMA3GFgYxdh4AAZ6n-ooSp8zL8KHoUCg4MAoGZoGYIY0GA9mAAzZh+PSQprQHB6BDQYDMXEAOhIVTmDUaHjhSw+K10XRoe0m21G4UZWx25jZBz4J3Olxu9wgj2erygIkhABsIHoaqhoGBglCIMBahAICgoN9fqCAUComqFHhgHAEULkaiFMTSZD5o0AHJy6kaQiremoLnM6ZAA) that I created to illustrate how to.

andrewhayward replied:
For the purposes of this PR, I wanted to leave both `aria-label` and `aria-labelledby` as optional props here, because of the number of components that use this and don't provide either; fixing all of those would dramatically increase the scope of the ticket. I created [a follow-up ticket to track that work](https://github.com/WordPress/gutenberg/issues/54055).

The desired mutual exclusivity could probably be clearer here though, and it may be worth adopting the `never` union syntax, even if they both remain optional...

```TypeScript
export type ColorPaletteProps = Pick< PaletteProps, 'onChange' > & {
  ...
} & (
  | {
      'aria-label'?: string;
      'aria-labelledby'?: never;
    }
  | {
      'aria-label'?: never;
      'aria-labelledby'?: string;
    }
);
```

ciampo replied:
Sounds good, we can do this as a follow-up

ciampo commented on packages/components/src/duotone-picker/duotone-picker.tsx:
```diff
@@ -120,8 +122,18 @@ function DuotonePicker( {
+	if ( ariaLabel ) {
+		ariaProps = { 'aria-label': ariaLabel as string };
+	} else if ( ariaLabelledby ) {
+		ariaProps = { 'aria-labelledby': ariaLabelledby as string };
```
Same as above re. the `as string` type cast

ciampo commented on packages/components/src/duotone-picker/types.ts:
```diff
@@ -39,6 +39,18 @@ export type DuotonePickerProps = {
+	 *
+	 * @todo Either this or `aria-label` should be required
+	 */
+	'aria-labelledby'?: string;
```
Same as above re. requiring one prop or the other

andrewhayward replied:
See [response above](#discussion_r1312998526).

ciampo commented on packages/components/src/gradient-picker/types.ts:
```diff
@@ -42,6 +42,18 @@ type GradientPickerBaseProps = {
 	 * @default 2
 	 */
 	headingLevel?: HeadingSize;
+	/**
+	 * A label to identify the purpose of the control.
+	 *
+	 * @todo Either this or `aria-labelledby` should be required
+	 */
+	'aria-label'?: string;
+	/**
+	 * An ID of an element to provide a label for the control.
+	 *
+	 * @todo Either this or `aria-label` should be required
+	 */
+	'aria-labelledby'?: string;
```
Same as above

andrewhayward replied:
See [response above](#discussion_r1312998526).

ciampo commented on packages/components/src/circular-option-picker/types.ts:
```diff
@@ -36,7 +40,20 @@ export type CircularOptionPickerProps = {
+	  }
+	| {
+			'aria-labelledby': string;
+	  }
```
Since these 2 props should probably be mutually exclusive, we could adopt the same approach suggested in a previous comment

andrewhayward replied:
This is me mixing up my Flow and TypeScript 🤦 Will fix.

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 -   Make the `Popover.Slot` optional and render popovers at the bottom of the document's body by default. ([#53889](https://github.com/WordPress/gutenberg/pull/53889)).
 -   `ProgressBar`: Add transition to determinate indicator ([#53877](https://github.com/WordPress/gutenberg/pull/53877)).
 -   Prevent nested `SlotFillProvider` from rendering ([#53940](https://github.com/WordPress/gutenberg/pull/53940)).
+-   Making Circular Option Picker a `listbox`. Note that while this changes some public API, new props are optional, and currently have default values; this will change in another patch. ([#52255](https://github.com/WordPress/gutenberg/pull/52255))
```
What are we planning on changing exactly in a future patch?

andrewhayward replied:
Sort of [touched on this above](#discussion_r1312998526), but ultimately all consumers should be providing their own label for the picker. `CircularOptionPicker` now requires an `aria-label` or `aria-labelledby`, and while `ColorPalette`, `DuotonePicker` and `GradientPicker` can all pass them through, they also provide a default value to obfuscate the need. Eventually they too will require one or the other, and not have a default; this will be addressed in #54055.

ciampo reviewed (commented):
Great job on this one so far, @andrewhayward !

I gave a quick look at Storybook and at the code and left some comments.

I'll have a closer look at how the component behaves in the editor in the next review round

Commit:
Removing redundant type coercion

Commit:
Enforcing mutual exclusivity of `-label`/`-labelledby`

Commit:
Making label exclusivity clearer

Commit:
Merge branch 'trunk' into 35292/improve_color_palette_keyboard_accessiblity--listbox

Commit:
Updating CHANGELOG.md

Commit:
Fixing broken test

ciampo reviewed (approved):
LGTM 🚀 

Code changes look good and test well as per instructions.

ciampo committed:
Merge branch 'trunk' into 35292/improve_color_palette_keyboard_accessiblity--listbox

ciampo commented:
Noticed https://github.com/WordPress/gutenberg/issues/54156 while testing, @andrewhayward this is also another good follow-up task

ciampo commented:
Added a dev note in the PR description
